### PR TITLE
Hollis/offline dashboard

### PR DESCRIPTION
**Pull request by hollisma**
# Description

- The other vision tools being built this summer (label annotating, label propagating, and retraining) can all be done with an isolated dataset. The dashboard currently requires an agent to be running, so this milestone is to let users annotate and retrain on a separate dataset not attached to an agent.
- Even though the offline dashboard does not use an agent on the backend, it still needs a backend to read/write to files and run label prop, so I created a dummy backend in ./server.py which can be run with `python server.py` .
- User flow
    - User clicks new "Offline" panel on right side, which prompts the user for a directory path. The directory needs three components: a depth folder, rgb folder, and a data.json mapping data point ids to the x, y, and yaw data. Each depth map is assumed to be in the `0000x.npy` format while rgb images are to be in the `0000x.jpg` format.
    - The first depth-image-pose instance will be displayed on the rgb, depth, and object dashboard feeds. Two buttons allow users to sequentially navigate through the instances. On any frame, users can draw annotations, which are then propagated to any other frame the user navigates to and saved to disk, assuming the new frame doesn't already have annotations.
    - Users can still save the annotations and retrain a model that can be found at the following filepath: `annotation_data/model/model_999.pth`

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [ ] Refactor
- [x] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [ ] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

## Before and After

New offline panel: 
image

Offline dashboard view: 
image

Offline dashboard start: 
Offline Dashboard Start

Navigation: 
Offline Dashboard Navigation

Label propagation: 
Offline Dashboard Label Prop

# Testing

Tested locally using files from `checkpoint/apratik/ActiveVision/active_vision/replica_random_exploration_data/apartment_0`. Made sure that the above flow works. 

To test, launch the pyro habitat in the docker container, run `python server.py` in the root folder, and run `yarn start` in dashboard/web. Navigate to the Offline panel on the right and enter the filepath to the data. 

# Limitations / Improvements

- I decided to only run label prop when the new frames do not already have any annotations because otherwise, backtracking would duplicate annotations. Users might want to still propagate labels on frames that have annotations, so we can add a toggle for this.
- In order to get to the final frame, users need to navigate through the frames one by one. Can include a navigation tool to jump to a specific frame.
- We should build support for videos in the future. A nice addition to accompany this would be to automatically go through the frames so that users don't need to manually go through each frame. Could be a toggle and done at a pace that the user sets.

# Checklist:

- [x] I have performed manual end-to-end testing of the feature in my environment.
- [x] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.

### CODE REVIEW

File: droidlet/dashboard/web/src/StateManager.js [hunk 1]
```diff
@@ -106,7 +108,7 @@ class StateManager {
     this.setTurkWorkerId(turkWorkerId);
 
     // set default url to actual ip:port
-    this.default_url = window.location.host;
+    // this.default_url = window.location.host;
```
Comment (anuragprat1k) [inline droidlet/dashboard/web/src/StateManager.js:111]: uncomment this line?

File: agents/locobot/label_prop.py [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+        save_coco(test_path, info, licenses, y_images, y_annots, categories)
+
+    # 2) Use train/test files to retrain detector
+    # Adapted from train_detector.ipynb
+    dataset_name = "annotation_coco"
+    image_dir = base_path + "rgb/"
+    train_data = dataset_name + "_train"
+    test_data = dataset_name + "_test"
+
+    if train_data in DatasetCatalog.list():
```
Comment (anuragprat1k) [inline agents/locobot/label_prop.py:278]: did `DatasetCatalog.clear()` and `MetadataCatalog.clear()` not work?

File: droidlet/dashboard/web/src/StateManager.js [hunk 3]
```diff
@@ -615,6 +663,102 @@ class StateManager {
     });
   }
 
+  goOffline(filepath) {
```
Comment (anuragprat1k) [inline droidlet/dashboard/web/src/StateManager.js:666]: can you add docstrings for each of these functions?

File: server.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        @sio.on("get_offline_frame")
+        def get_offline_frame(sid, data): 
+            rgb, depth = LP.get_offline_frame(data)
+            sio.emit("rgb", rgb)
+            sio.emit("depth", depth)
+
+
+if __name__ == "__main__":
+    
+    base_path = os.path.dirname(__file__)
```
Comment (anuragprat1k) [inline server.py:154]: how is this server kicked off while the dashboard is running?

File: server.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+)
+
+class OfflineInstance(): 
+
+    def __init__(self): 
+        self.init_event_handlers()
+    
+    def init_event_handlers(self): 
+        @sio.on("offline_label_propagation")
+        def offline_label_propagation(sid, postData):
```
Comment (anuragprat1k) [inline server.py:41]: how much of this code is a duplication of the online label prop? can we reduce code duplication here?

**Review by anuragprat1k — CHANGES_REQUESTED** [review]

File: agents/locobot/label_prop.py (hunk 2 above)
Comment (hollisma) [reply agents/locobot/label_prop.py:278]: I put these changes in the hollis/fix_retraining branch but I can include them here as well

File: server.py (hunk 4 above)
Comment (hollisma) [reply server.py:154]: Server.py is the backend for the offline dashboard and is the agent-free replacement for locobot_agent.py

File: server.py (hunk 5 above)
Comment (hollisma) [reply server.py:41]: The differences are in how we read the input rgb/seg data, which account for about a third of the lines in the offline_label_propagation and offline_save_rgb_seg functions. I'll replace the duplicate code and refactor it to the label_prop file.

**Review by anuragprat1k — APPROVED** [review]

**Pull request merged**